src/libclipper/src/query_processor.cpp

```diff
@@ -94,21 +101,29 @@ boost::future<Response> QueryProcessor::predict(Query query) {
   boost::promise<Response> response_promise;
   auto response_future = response_promise.get_future();
 
-  // NOTE: We capture the num_completed and completed_flag variables
+  // NOTE: We capture the num_completed, completed_flag, and default_explanation
+  // variables
   // so that they outlive the composed_futures.
   response_ready_future.then([
     this, query, query_id, response_promise = std::move(response_promise),
     selection_state, current_policy, task_futures = std::move(task_futures),
-    num_completed, completed_flag
+    num_completed, completed_flag, default_explanation
   ](auto) mutable {
 
     vector<Output> outputs;
     vector<VersionedModelId> used_models;
+    bool all_tasks_timed_out = true;
     for (auto r = task_futures.begin(); r != task_futures.end(); ++r) {
       if ((*r).is_ready()) {
         outputs.push_back((*r).get());
+        all_tasks_timed_out = false;
       }
     }
+    if (all_tasks_timed_out && !task_futures.empty() && !default_explanation) {
+      default_explanation =
+          "Failed to retrieve a prediction response within the specified "
+          "latency SLO";
+    }
```

Comment:
Instead of throwing and catching a `NoModelsFoundError`, can you just check if `task_futures.empty()` here and use that to set the `default_explanation`? Also, set the default explanation to "No connected models found for query" instead of "No registered models found for query", because it's possible for a model to be registered with Clipper but no containers connected yet and the user will still see this message.